crates/ide/src/ide/assists/rewrite_string.rs

```diff
@@ -0,0 +1,322 @@
+                "Replace the double quoted string with an indented string",
+                AssistKind::QuickFix,
+                vec![TextEdit {
+                    delete: node.syntax().text_range(),
+                    insert: text.into(),
+                }],
+            );
+
+            Some(())
+        }
+    } else if let Some(node) = ctx.covering_node::<ast::IndentString>() {
+        let mut text = String::from('"');
+        let _ = strip_indent::<Infallible>(&node, |part| {
+            match part {
+                StrippedStringPart::Fragment(frag) => {
+                    escape_dquote_string(&mut text, frag);
+                }
+                StrippedStringPart::Escape(esc) => {
+                    escape_dquote_string(&mut text, unescape_string_escape(esc.text()));
+                }
+                StrippedStringPart::Dynamic(dyna) => {
+                    text += &dyna.syntax().to_string();
+                }
+            }
+            Ok(())
+        });
+        text.push('"');
+
+        ctx.add(
+            "rewrite_indented_string",
+            "Replace the indented string with a double quoted string",
+            AssistKind::QuickFix,
+            vec![TextEdit {
+                delete: node.syntax().text_range(),
+                insert: text.into(),
+            }],
+        );
+
+        Some(())
+    } else {
+        None
+    }
+}
+
+/// Escape `string` and write to `out`,
+/// `string` may be only a fragment of the original Nix string.
+fn escape_dquote_string(out: &mut String, text: &str) {
+    let mut xs = text.chars();
+    while let Some(x) = xs.next() {
+        match x {
+            '"' | '\\' => {
+                out.push('\\');
+                out.push(x);
+            }
+            '$' => match xs.next() {
+                Some('{') => out.push_str("\\${"),
+                Some(y) => {
+                    out.push('$');
+                    out.push(y);
+                }
+                // It's impossible to know from this context whether the next
+                // character is '{' or not, so we assume it is just to be safe
+                None => out.push_str("\\$"),
+            },
+            '\n' => out.push_str("\\n"),
+            '\r' => out.push_str("\\r"),
+            '\t' => out.push_str("\\t"),
+            _ => out.push(x),
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use expect_test::expect;
+
+    define_check_assist!(super::rewrite_string);
+
+    #[test]
+    fn uri() {
+        check("$0https://nixos.org", expect![r#""https://nixos.org""#]);
+        check(
+            "https://github.com/oxalica/nil$0",
+            expect![r#""https://github.com/oxalica/nil""#],
+        );
+    }
+
+    #[test]
+    fn quote_attr() {
+        check("{ $0foo = bar; }", expect![r#"{ "foo" = bar; }"#]);
+        check("{ f$0oo.bar = baz; }", expect![r#"{ "foo".bar = baz; }"#]);
+        check("{ foo.bar$0 = baz; }", expect![r#"{ foo."bar" = baz; }"#]);
```

Comment:
These 3 samples seem to be duplicates to me. Do they cover anything more than, say just the second one?

So do tests below.

Reply:
Removed a few tests